### DISCUSSION TITLE
p.248 リスト20.28の`if err = u.ComparePassword(pw); err != nil {`について

### DISCUSSION THREAD

**Discussion by youta32449999** [budougumi0617/go_todo_app#59] [Q&A]
p.248 リスト20.28内の`if err = u.ComparePassword(pw); err != nil {`というコードで`:=`ではなく`=`を使用しているのはなぜでしょうか?
浅学で大変申し訳ないのですが、ご教授いただければ幸いです。どうぞよろしくお願いいたします。
```go
// service/login.go

func (l *Login) Login(ctx context.Context, name, pw string) (string, error) {
	u, err := l.Repo.GetUser(ctx, l.DB, name)
	if err != nil {
		return "", fmt.Errorf("failed to list: %w", err)
	}
	if err = u.ComparePassword(pw); err != nil {
		return "", fmt.Errorf("wrong password: %w", err)
	}
	jwt, err := l.TokenGenerator.GenerateToken(ctx, *u)
	if err != nil {
		return "", fmt.Errorf("failed to generate JWT: %w", err)
	}

	return string(jwt), nil
}
```

**Comment by budougumi0617** [comment]
ご指摘ありがとうございます。これは「`=` を使用できるけど、使用してはいけない」コードですね。
コードと正誤表を更新しました。
`if` 文の条件の前にステートメントで変数を宣言するのは「`if` 文のスコープの中のみで変数を有効にすること」にメリットがあるからですが、これでは意味がないですね…。
`GetUser` メソッドの戻り値で `err` 変数を宣言しているからこのように書けますが、コードとしてはよろしくない状態でした。ありがとうございます。

https://github.com/budougumi0617/go_todo_app/releases/tag/v1.0.7

**Reply by youta32449999** [reply]
ご回答ありがとうございます!
>if 文の条件の前にステートメントで変数を宣言するのは「if 文のスコープの中のみで変数を有効にすること」にメリットがある

`error`のみを返す関数またはメソッドの呼び出しがif文の条件前にいる形になっているのは、このようなメリットがあるからなのですね!

こちらの本のおかげで特にcontextとは?のようなモヤモヤが取り除かれて、Goのコードや他の本がとても読みやすくなりました。とても良い本を執筆いただきありがとうございます!
